### PR TITLE
✅ Add initial integration test of EVM, Host, and State

### PR DESCRIPTION
**Pull request by sheapwilliams**
More to follow, but I want to publish this to get some eyes on it.

Future work will look into getting GUnit to work with ctest. If I can, I'll refactor this before working on the TODO tests.

### CODE REVIEW

File: cmake/test.cmake [hunk 1]
```diff
@@ -7,19 +7,35 @@ function(add_unit_test)
...
+        ${ADD_UNIT_TEST_TARGET} PUBLIC GTest::GTest monad_unit_test_common
+                                       ${ADD_UNIT_TEST_LIBRARIES}
     )
     gtest_discover_tests(${ADD_UNIT_TEST_TARGET})
 endfunction()
+
+function(add_integration_test)
+    set(ONE_VALUE_ARGS TARGET)
+    cmake_parse_arguments(ADD_INT_TEST "" "${ONE_VALUE_ARGS}" "" ${ARGN})
+    message(STATUS "int target is: ${ADD_INT_TEST_TARGET}")
```
Comment (zander-xyz) [inline cmake/test.cmake]: Seems unnecessary, but if you want to keep this log around, can we spell out integration?

File: cmake/test.cmake [hunk 2]
```diff
@@ -7,19 +7,35 @@ function(add_unit_test)
...
+
+    add_unit_test(${ADD_INT_TEST_TARGET} ${ARGN})
+    target_link_libraries(${ADD_INT_TEST_TARGET} PUBLIC monad)
+
+    add_custom_target(integration-tests)
```
Comment (zander-xyz) [inline cmake/test.cmake]: this fails if we have more than one integration test, I think you ought to check if integration-tests exists first before adding it

File: include/monad/db/state.hpp [hunk 3]
```diff
@@ -149,6 +149,13 @@ struct State
             code_.revert();
         }
 
+        [[nodiscard]] bytes32_t get_block_hash(int64_t number) const noexcept
```
Comment (zander-xyz) [inline include/monad/db/state.hpp:152]: yeah it really sucks that the evm host API takes block number as signed...i feel your pain

what do you think about pushing the signed-ness up to the evmc_host layer and doing the translation to unsigned there?

File: cmake/test.cmake (hunk 1 above)
Comment (sheapwilliams) [reply cmake/test.cmake]: Nope, I'll remove it.

File: include/monad/db/state.hpp (hunk 3 above)
Comment (sheapwilliams) [reply include/monad/db/state.hpp:152]: Not the worst impulse, but I'd really like to leave the evmc host as simple composition.  Implementors can deal with the details.

Comment (zander-xyz) [reply include/monad/db/state.hpp:152]: yeah i guess there are two schools of thought
1. use the the evmc_host as a layer to hide-away the warts of the interface itself
2. what you said

either is fine, your call

File: test/integration/evm_state_host/evm_state_host.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    0x000000000000000000000000000000000000000000000000000000000000004c_bytes32;
+
+using code_db_t = std::unordered_map<address_t, byte_string>;
+using account_store_db_t = db::InMemoryDB;
+
+template <class TState, concepts::fork_traits<TState> TTraits>
+using traits_templated_static_precompiles_t = execution::StaticPrecompiles<
+    TState, TTraits, typename TTraits::static_precompiles_t>;
+
+TEST(EvmInterpStateHost, return_existing_storage)
```
Comment (zander-xyz) [inline test/integration/evm_state_host/evm_state_host.cpp:43]: thoughts on making this a typed test and doing it on all types of DBs? similar to the *store unit tests

File: include/monad/db/state.hpp (hunk 3 above)
Comment (sheapwilliams) [reply include/monad/db/state.hpp:152]: Yea - I think this is the right idea for now.  If we find ourselves with a different composition and copying the signed->unsigned code, then we can fix it then.  If that different composition doesn't have that (cause reasons) then we got it right.  We can figure it then.

File: test/integration/evm_state_host/evm_state_host.cpp (hunk 4 above)
Comment (sheapwilliams) [reply test/integration/evm_state_host/evm_state_host.cpp:43]: I'm not totally allergic to it, but permutating all the genetic possibilities of our integration and test plan can get out of hand, right?  I think the backend DB is a detail that the DBInterface hides away for this test - we are really just using the in-memory for testing, and integrating the evm, host, and state.  The back-end integration between DBInterface is taken care of by other tests.  They will come together at the next higher level, probably with the block playback.

So, I'm comfortable with only these types tested at this level of testing, esp. cause the trie building is the slowest part of our testing.

This is a subjective thing for sure to get the test coverage right while keeping pipelines and job durations reasonable.  Hit me up on zoom if you want to talk more about it.

Comment (zander-xyz) [reply test/integration/evm_state_host/evm_state_host.cpp:43]: One alternative viewpoint could be formed by examining what this unit test is doing. This unit test is storing a key value pair into the storage of an account, then reading from the storage of that account, storing that into "memory", then returning it.

That involves the implementation of the DB interface to be correct. Contrast that to a fake, where you really don't care what the implementation is (in which case, I wouldn't care what backend you used). Here, you *really do care*. You care that the implementation is right. You want and demand it to be right through your assertions. And in our case, gtest makes it super easy for us to exercise both the "easy" case (InMemoryDb), and harder cases (*TrieDbs). 

Does this make sense?

Comment (sheapwilliams) [reply test/integration/evm_state_host/evm_state_host.cpp:43]: If I worked on the DB, I might frame things around that POV.  But I didn't, and I don't.  Let's zoom if you want to chat test coverage.  It is subjective and not always easy to get right, so it is worth a discussion.

**Pull request merged**